crates/multi-transaction-payment/src/tests.rs

```diff
@@ -0,0 +1,417 @@
+        })
+        .unwrap()
+    }
+
+    pub(super) fn setup_dex(currency_1: CurrencyId, currency_2: CurrencyId) {
+        let lp = 2323;
+        assert_ok!(<Test as dex_general::Config>::MultiCurrency::deposit(
+            currency_1,
+            &lp,
+            1000000000000000000000
+        ));
+        assert_ok!(<Test as dex_general::Config>::MultiCurrency::deposit(
+            currency_2,
+            &lp,
+            1000000000000000000000
+        ));
+
+        assert_ok!(DexGeneral::create_pair(
+            RuntimeOrigin::root(),
+            currency_1,
+            currency_2,
+            15,
+        ));
+
+        assert_ok!(DexGeneral::add_liquidity(
+            RuntimeOrigin::signed(lp),
+            currency_1,
+            currency_2,
+            1000000000000000000000,
+            1000000000000000000000,
+            0,
+            0,
+            u64::MAX
+        ));
+    }
+    #[derive(Debug, PartialEq)]
+    pub(super) struct BalanceChange {
+        pub(super) native_cost: i128,
+        pub(super) fee_cost: i128,
+    }
+
+    pub(super) fn dry_run_given_call(
+        who: u128,
+        native_currency: CurrencyId,
+        fee_currency: CurrencyId,
+        pays_fee: Pays,
+        call: RuntimeCall,
+        tip: u128,
+    ) -> BalanceChange {
+        dry_run(|| {
+            let initial_fee_balance = 1000000000000;
+            assert_ok!(Tokens::set_balance(
+                RuntimeOrigin::root(),
+                who,
+                fee_currency,
+                initial_fee_balance,
+                0
+            ));
+            let initial_native_balance = Tokens::free_balance(native_currency, &who);
+
+            let payment_extension = pallet_transaction_payment::ChargeTransactionPayment::<Test>::from(tip);
+
+            let info = call.get_dispatch_info();
+            let len = 1231;
+
+            // check that validate executes successfully, but don't commit storage
+            dry_run(|| {
+                payment_extension.validate(&who, &call, &info, len).unwrap();
+                if let Pays::Yes = pays_fee {
+                    assert!(Tokens::free_balance(fee_currency, &who) < initial_fee_balance);
+                }
+            });
+
+            let pre = payment_extension.pre_dispatch(&who, &call, &info, len).unwrap();
+            if let Pays::Yes = pays_fee {
+                assert!(Tokens::free_balance(fee_currency, &who) < initial_fee_balance);
+            }
+
+            let result = call.dispatch(RuntimeOrigin::signed(who));
+            let post_info = match result {
+                Ok(post_info) => post_info,
+                Err(err) => err.post_info,
+            };
+
+            Payment::post_dispatch(Some(pre), &info, &post_info, len, &Ok(())).unwrap();
+
+            BalanceChange {
+                native_cost: (initial_native_balance as i128) - (Tokens::free_balance(native_currency, &who) as i128),
```

Comment:
I would expect this value to always be less than or equal to zero, if a correct amount of `fee_balance` is converted into the native currency

Reply:
This function makes no assumption on the fee currency - it can be the native token as well (i.e. without swapping), meaning it can be positive or negative. That's why it's i128 rather than u128